packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/TypeConverterProvider.kt

```diff
@@ -0,0 +1,107 @@
+@file:OptIn(ExperimentalStdlibApi::class)
+
+package expo.modules.kotlin.types
+
+import com.facebook.react.bridge.Dynamic
+import com.facebook.react.bridge.ReadableArray
+import com.facebook.react.bridge.ReadableMap
+import expo.modules.kotlin.exception.MissingTypeConverter
+import expo.modules.kotlin.records.Record
+import expo.modules.kotlin.records.RecordTypeConverter
+import kotlin.reflect.KClass
+import kotlin.reflect.KType
+import kotlin.reflect.full.createType
+import kotlin.reflect.full.isSubclassOf
+import kotlin.reflect.typeOf
+
+interface TypeConverterProvider {
+  fun obtainTypeConverter(type: KType): TypeConverter<*>
+}
+
+inline fun <reified T : Any> obtainTypeConverter(): TypeConverter<T> {
+  @Suppress("UNCHECKED_CAST")
+  return TypeConverterProviderImpl.obtainTypeConverter(typeOf<T>()) as TypeConverter<T>
+}
+
+inline fun <reified T> convert(value: Dynamic): T {
+  val converter = TypeConverterProviderImpl.obtainTypeConverter(typeOf<T>())
+  return converter.convert(value) as T
+}
+
+fun convert(value: Dynamic, type: KType): Any? {
+  val converter = TypeConverterProviderImpl.obtainTypeConverter(type)
+  return converter.convert(value)
+}
+
+object TypeConverterProviderImpl : TypeConverterProvider {
+  private val cachedConverters = createCashedConverters(false) + createCashedConverters(true)
+
+  override fun obtainTypeConverter(type: KType): TypeConverter<*> {
+    cachedConverters[type]?.let {
+      return it
+    }
```

Comment:
I read the PR's description, but I still don't get it. Why do you cache only some types here, while others are returned later in this function each time it's called?

Reply:
Ok, it isn't the code that I referred to in the description. The type converter is cached in the `AnyType` class. 

However, for some basic types like we don't have to create a multiple converter object for each prop, cause the conversion looks always the same. But, if we are talking about conversion to list for instance. We can't just create one converter. Conversion to the list of int will be different than the list of strings. 

So to sum everything up, for basic types, we share the same instance of converter across the whole application. For more complicated types, we create a dedicated converter for each prop. After that, we store the reference to the converter in AnyType. So we don't have to obtain a converter before each call.

Reply:
Yeah, that makes sense, thanks for the explanation